File: Examples/DevCycle-iOS-Example-App-ObjC/DevCycle-iOS-Example-App-ObjC/ViewController.m
//
//  ViewController.m
//  DevCycle-Example-App-ObjC
//
//

#import "ViewController.h"
#import "DevCycleManager.h"
@import DevCycle;

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UIButton *loginButton;
@property (atomic) DevCycleClient *client;
@property BOOL loggedIn;

@end

@implementation ViewController

- (IBAction)loginButtonPressed:(id)sender {
    __weak typeof(self) weakSelf = self;
    if (self.loggedIn) {
        [self.client resetUser:^(NSError *error, NSDictionary<NSString *,id> *variables) {
            NSLog(@"DevCycle: Reset User!");
            NSLog(@"%@", variables);
            weakSelf.loggedIn = NO;
            [weakSelf.loginButton setTitle:@"Log In" forState:UIControlStateNormal];
        }];
    } else {
        DevCycleUser *user = [DevCycleUser initializeWithUserId:@"my-user"];
        user.userId = @"my-user";
        user.name = @"My Name";
        user.language = @"en";
        user.country = @"CA";
        user.email = @"my@email.com";
        
        [self.client identifyUser:user callback:^(NSError *error, NSDictionary<NSString *,id> *variables) {
            if (error) {
                return NSLog(@"Error calling DevCycleClient identifyUser:callback: %@", error);
            }
            NSLog(@"Identified User!");
            NSLog(@"%@", variables);
            weakSelf.loggedIn = YES;
            [weakSelf.loginButton setTitle:@"Log Out" forState:UIControlStateNormal];
        }];
    }
}

- (IBAction)track:(id)sender {
    NSError *err = nil;
    DevCycleEvent *event = [DevCycleEvent initializeWithType:@"my-event"];
    [self.client track:event err:&err];
    if (err) {
        NSLog(@"Error calling DevCycleClient track:err: %@", err);
    } else {
        NSLog(@"Tracked event to DevCycle");
    }
}


- (IBAction)logAllFeatures:(id)sender {
    NSLog(@"All Features: %@", [self.client allFeatures]);
    NSLog(@"All Variables: %@", [self.client allVariables]);
}


- (void)viewDidLoad {
    [super viewDidLoad];
    self.client = [[DevCycleManager sharedManager] client];
}


@end

File: Examples/DevCycle-iOS-Example-App-Swift/DevCycle-iOS-Example-App-Swift/ViewController.swift
//
//  ViewController.swift
//  DevCycle-Example-App
//
//

import DevCycle
import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var titleHeader: UILabel!
    @IBOutlet weak var loginButton: UIButton!

    var loggedIn: Bool = false
    var titleHeaderVar: DVCVariable<String>?
    var loginCtaVar: DVCVariable<String>?

    override func viewDidLoad() {
        super.viewDidLoad()
        Task {
            guard let client = await DevCycleManager.shared.clientAsync else { return }
            self.loginCtaVar = client.variable(key: "login-cta-copy", defaultValue: "Log")
                .onUpdate(handler: { value in
                    self.setLoginButtonTitle(self.loggedIn)
                })
            self.titleHeaderVar = client.variable(
                key: "title-header-copy", defaultValue: "DevCycle iOS Example App"
            ).onUpdate(handler: { value in
                self.setTitleHeader()
            })
            self.setTitleHeader()
            self.setLoginButtonTitle(false)
        }
    }

    func setTitleHeader() {
        guard let titleHeaderVar = self.titleHeaderVar else {
            return
        }
        self.titleHeader.text = titleHeaderVar.value
    }

    func setLoginButtonTitle(_ bool: Bool) {
        guard let loginCta = self.loginCtaVar else {
            return
        }
        self.loggedIn = bool
        self.loginButton.setTitle("\(loginCta.value) \(self.loggedIn ? "out" : "in")", for: .normal)
    }

    @IBAction func loginButtonPressed(_ sender: Any) {
        Task {
            guard let client = await DevCycleManager.shared.clientAsync else { return }
            if self.loggedIn {
                do {
                    _ = try await client.resetUser()
                    self.setLoginButtonTitle(false)
                    print("Reset User!")
                } catch {
                    print("Error resetting user: \(error)")
                }
            } else {
                do {
                    let user = try DevCycleUser.builder()
                        .userId("my-user1")
                        .email("my-email@email.com")
                        .country("CA")
                        .name("Ash Ketchum")
                        .language("EN")
                        .customData([
                            "customkey": "customValue"
                        ])
                        .privateCustomData([
                            "customkey2": "customValue2"
                        ])
                        .build()
                    _ = try await client.identifyUser(user: user)
                    self.setLoginButtonTitle(true)
                    print("Logged in as User: \(String(describing: user.userId))!")

                    let numKeyValue = client.variableValue(key: "num_key", defaultValue: Double(0))
                    print("Num_key is \(numKeyValue)!")

                    let numKeyDefaultedValue = client.variableValue(
                        key: "num_key_defaulted", defaultValue: Double(0))
                    print("Num_key_defaulted is \(numKeyDefaultedValue)!")
                } catch {
                    print("Error identifying user: \(error)")
                }
            }
        }
    }

    @IBAction func track(_ sender: Any) {
        Task {
            guard let client = await DevCycleManager.shared.clientAsync else { return }
            let event = try! DevCycleEvent.builder()
                .type("my_event")
                .target("my_target")
                .value(3)
                .metaData(["key": "value"])
                .clientDate(Date())
                .build()
            client.track(event)
            print("Tracked event to DevCycle")
        }
    }

    @IBAction func logAllFeatures(_ sender: Any) {
        Task {
            guard let client = await DevCycleManager.shared.clientAsync else { return }
            print("All Features: \(client.allFeatures())")
            print("All Variables: \(client.allVariables())")
        }
    }
}
